### DISCUSSION TITLE
`spot_price` definition in the input csv files

### DISCUSSION THREAD

**Discussion by topherbuckley** [eprbell/dali-rp2#129] [Q&A]
According to the docs the `spot_price` is defined as:

>(optional): value of 1 unit of the given cryptocurrency at the time the transaction occurred; If the value is unavailable, to direct DaLI to read it from Internet historical data, write in __unknown and use the -s command line switch;

I would guess this is meant to be the value in the paired coin/currency, but it is a bit find ambiguous so wanted to check. For example lets say I have a BUY line on a ETHBTC exchange, where I'm buying ETH and paying in BTC, I'd assume spot_price should be how much 1 ETH is worth in BTC, or is it perhaps how much 1 BTC is worth in ETH? With the column called `fiat_out_no_fee`, I'd assume `spot_price` in this case wouldn't have anything to do with fiat, but looking at the definition for the fees:

>crypto_fee: (optional) crypto value of the transaction fees. This field is mutually exclusive with fiat_fee. If fiat_fee is assigned, crypto_fee is set to 0. If crypto_fee is assigned, fiat_fee is set to crypto_fee * spot_price. The reason for this behavior is that if the fee is paid in fiat, then no crypto is used for the fee, but if the fee is paid in crypto, then its converted fiat value is needed to compute taxes. Note that RP2 models a non-zero crypto_fee with a separate fee-typed out-transaction.

I'd say `spot_price` is supposed to be the fiat equivalent (i.e. in the first example, how much 1 ETH is worth in USD or JPY). If this is the case, then which fiat is supposed to be used or does it not matter so long as you always use one? I've got transactions from both USD and JPY accounts, so I'm not quite sure how this is supposed to be handled. Should I be looking up historical exchange rates myself and converting them all to one currency or can dali-rp2 handle this somehow?

Also worth noting is that `spot_price` also appears in the `intra_csv_file`. If I'm moving ETH from one account to another, what then does `spot_price` represent? Here again I'd guess it is some fiat equivalent.

**Comment by macanudo527** [comment]
The `spot_price` should be the price of the digital asset in fiat, which defaults to USD. You can declare a different fiat by passing in the `fiat_code` of it. If it is not easily obtainable, I recommend letting the CCXT pair converter handle looking up the price for you. 

The CCXT pair converter will 'route' the price to fiat. It will try to prioritize the exchange you made the purchase/sale at, but if that is not available will use the default exchange (currently Kraken). 

For example, if you make a trade for ETH in BTC. It will look up the price for ETH in BTC at that moment if the market exists on that exchange. Then find the price of BTC in the fiat you are using. 

ETH -> BTC -> USD

The final fiat is determined by what version of Dali-rp2 you are using. For Dali_us it is USD and Dali_jp it is JPY. so if you are filing taxes in Japan and America, you will need to run the program twice with the same data but different entry points (dali_us and dali_jp). 

I hope that makes sense.

**Reply by topherbuckley** [reply]
Thank you! That all makes sense other than I can't find code or docs regarding the `fiat_code` you mentioned. Is that another column in the input csv files, or a terminal argument? Happy to read code or docs, so if you can point me towards where I can read about that, I think I'd be all set. Cheers!

**Reply by macanudo527** [reply]
Sorry, it doesn't look like it is in the docs. I'll put it on my to-do list for when I revise the Bitbank/Binance typo. It's actually called `fiat_ticker`. Here is an example in the [CCXT abstract plugin](https://github.com/eprbell/dali-rp2/blob/91a473119f32b416609bc4e88214436216d0e588/src/dali/abstract_ccxt_input_plugin.py#L547-L566). It needs to be the 3 letter ISO code.

**Reply by topherbuckley** [reply]
Thanks! I'll have to read through the CCXT abstract plugin to get a better handle on this. Maybe I'm reading from an irrelevant file, but I was assuming [manual.py](https://github.com/eprbell/dali-rp2/blob/91a473119f32b416609bc4e88214436216d0e588/src/dali/plugin/input/csv/manual.py) was governing the import of manual csv imports. If that is the case, it looks like the columns are ordered args rather than keyword args from what I'm reading [here](https://github.com/eprbell/dali-rp2/blob/91a473119f32b416609bc4e88214436216d0e588/src/dali/plugin/input/csv/manual.py#L137-L151) so then how would this `fiat_ticker` info get past this point? Is `fiat_ticker` only supported in certain plugins?

**Reply by macanudo527** [reply]
The manual.py is out of date. Sorry, I'll try to update it here soon. All plugins support it because it is a property of `InTransaction` and `OutTransaction` not limited by the plugins.

**Reply by topherbuckley** [reply]
I think I'm misunderstanding something. Using the test files in the README as an example:
- [test_config.ini](https://raw.githubusercontent.com/eprbell/dali-rp2/main/config/test_config.ini)
- [test_manual_in.csv](https://raw.githubusercontent.com/eprbell/dali-rp2/main/input/test_manual_in.csv)

I still am not sure if `fiat_ticker` is supposed to be 

1. a column header in the input csv files (i.e. after the Notes column)
2. a parameter passed to a data loader e.g within `test_config.ini`:
```
[dali.plugin.input.csv.manual]
fiat_ticker = JPY
in_csv_file = input/test_manual_in.csv
```
3. Something else

I tested 1. and 2. and didn't notice any effect to the output.  Can you help me understand where specifically to use `fiat_ticker` relative to these example files?

**Reply by topherbuckley** [reply]
Thinking this through, I think it might just be prudent to use "__unknown" and ignore the spot price data I have in JPY and only include known spot price when it is in USD. I assume this will work, but still curious if there is a way to make use of the data I have as I assume it is more accurate for the transactions in question.

**Reply by eprbell** [reply]
I checked the code and found out why you're not seeing any effect. It's because the manual plugin has not been updated to use `fiat_ticker`. For example, here's how the manual plugin instantiates an InTransaction:
https://github.com/eprbell/dali-rp2/blob/main/src/dali/plugin/input/csv/manual.py#L136

You can see that `fiat_ticker` is missing and so it always gets assigned the default value (which is the native fiat). The same is true of Out/IntraTransaction. It should be easy to add support for it though, if you'd like to submit a PR.

Then the transaction resolver will be able to identify transactions with a fiat ticker that is different than the native fiat (USD for dali_us and JPY for dali_jp) and convert them to native fiat: see [`_convert_fiat_fields_to_native_fiat()`](https://github.com/eprbell/dali-rp2/blob/main/src/dali/transaction_resolver.py#L241)

Hope this helps.

**Reply by topherbuckley** [reply]
Thanks! Yeah, I think that makes sense. Is there any desire to move towards kwargs instead of an ordered list? I feel that might be less prone to silent bugs (e.g. headers in csv file in the wrong order). Or is there some error handling somewhere I haven't noticed that could detect such things? 

Also I guess I'd have to add it at the end of the existing ordered list, else cause backward compatibility issues and make everyone's existing csv files import incorrectly. I can see this becoming a bit of a mess if things keep getting added in chronological order rather than putting `fiat_ticker` next to the fiat amount (i.e. `spot_price`). 

Anyways, I'll go ahead and test it out as the final element in the existing ordered list.

**Reply by topherbuckley** [reply]
While putting in the edits, I noticed some of the INTRA fields are null checked and replaced with `Keyword.UNKNOWN.value` while similar fields in IN/OUT are not (e.g. [here](https://github.com/eprbell/dali-rp2/blob/815e6f161372f40ea0684cd6f6d944512a6ab17c/src/dali/plugin/input/csv/manual.py#L242) vs [here](https://github.com/eprbell/dali-rp2/blob/main/src/dali/plugin/input/csv/manual.py#L191). Is there are reason for this or should is the null check elsewhere for In/out transactions fields?

**Reply by topherbuckley** [reply]
Adding untested PR [here](https://github.com/eprbell/dali-rp2/pull/134). I mention in the comments of the PR that I'm having issues with the build/test workflow, so if you'd prefer, we can continue to discuss on the PR and I can make updates as necessary.

**Reply by eprbell** [reply]
The best way to fix this would be to add field index qualifiers to the manual plugin (similar to what [RP2 does](https://github.com/eprbell/rp2/blob/main/src/rp2/configuration.py#L239) with the `in_header`, `out_header` and `intra_header` sections of its [configuration file](https://github.com/eprbell/rp2/blob/main/docs/input_files.md#the-config-file). This would allow the user to select their custom format. Absent that, yes, we'll have to add `fiat_ticker` at the end to preserve backward compatibility.

The `UNKNOWN` values in intra-transactions are used only if the [transaction is partial](https://github.com/eprbell/dali-rp2/blob/main/docs/configuration_file.md#partial-transactions-and-transaction-resolution).

Sure, let's continue the discussion in the PR: it'll be easier.

**Reply by topherbuckley** [reply]
I had another comment on this unrelated to the PR. Is there any way of making use of a spot price given by an exchange for crypto pairs? For example, I get the "traded price" from bitflyer for a ETH/BTC buy transaction which is given as the value of 1 ETH in BTC used at the time of trade. If the `spot_price` in the input csv is only intended for fiat price, I guess my recommended course of action would be to use `__unknown` for all crypto2crypto transactions, and only use the `spot_price` for crypto2fiat transactions. Is that correct?

**Reply by macanudo527** [reply]
Not really, the `spot_price` is always fiat since that is what is important for taxes. A crypto to crypto trade is actually a conversion. RP2 will model this as a sale of one of the assets into fiat, then the purchase of the other asset will use that amount of fiat as the basis cost. This is how pretty much all tax agencies want you to record the transaction. 

So, let's take the ETH/BTC example.

You trade 0.1 BTC for 1 ETH. 
Dali (and your tax agency) will calculate this as:
 sale of 0.1 BTC for 3000 USD (if BTC is at $30,000)
 purchase of 1 ETH for 3000 USD (that becomes your basis cost if you sell it later).

So, the "spot price" of ETH in BTC is actually not really needed in terms of taxes. A sale of ETH for BTC is actually two transactions in Dali and for taxes.

I hope that makes sense.

**Reply by topherbuckley** [reply]
Yep, that makes sense. I was aware of that, but just thought making use of the data might improve accuracy in some way but regardless, we'd have to look up the fiat price so I suppose not. Thanks for confirming!
